enumerables.rb

```diff
@@ -0,0 +1,101 @@
+require_relative 'helpers'
+module Enumerable
+  def my_each
+    return enum_for(:my_each) unless block_given?
+
+    array = *self
+    array.size.times { |index| yield(array[index]) }
+    self
+  end
+
+  def my_each_index
+    return enum_for(:my_each) unless block_given?
+
+    array = *self
+    array.size.times { |index| yield(index) }
+  end
+
+  def my_select
+    return enum_for(:my_each) unless block_given?
+
+    array = *self
+    new_array = []
+    array.my_each { |item| new_array.push(item) if yield(item) }
+    new_array
+  end
+
+  def my_all?(*patt)
+    array = *self
+    return Helper.given_patt_for_all(array, patt[0]) unless patt.empty?
+
+    if block_given?
+      array.my_each { |item| return false unless yield(item) }
+    else
+      array.my_each { |item| return false unless item }
+    end
+    true
+  end
+
+  def my_any?(*patt)
+    array = *self
+    return Helper.given_patt_for_any(array, patt[0]) unless patt.empty?
+
+    if block_given?
+      array.my_each { |item| return true if yield(item) }
+    else
+      array.my_each { |item| return true if item }
+    end
+    false
+  end
+
+  def my_none?(*patt)
+    array = *self
+    return Helper.given_patt_for_none(array, patt[0]) unless patt.empty?
+
+    if block_given?
+      array.my_each { |item| return false if yield(item) }
+    else
+      array.my_each { |item| return false if item }
+    end
+    true
+  end
+
+  def my_count(*value)
+    array = *self
+    count = 0
+
+    if block_given?
+      array.my_each { |item| count += 1 if yield(item) }
+      return count
+    elsif value.size.positive?
+      array.my_each { |item| count += 1 if item == value[0] }
+      return count
+    end
+
+    array.size
+  end
+
+  def my_map
+    return enum_for(:my_each) unless block_given?
+
+    arr = *self
+    new_array = []
+    arr.my_each { |item| new_array << yield(item) }
+    new_array
+  end
```

Comment:
* [ ] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_map 

- [ ] returns an Enumerator if no block is given
```ruby
[2,4,7,11].my_map should be equal to [2,4,7,11].map #<Enumerator: [2, 4, 7, 11]:map
```
- Yours is always expecting for a block
![image](https://user-images.githubusercontent.com/52209067/78953617-b9da5c80-7a96-11ea-9978-0fc69e5fb8d6.png)

- Real map
![image](https://user-images.githubusercontent.com/52209067/77697646-dc3b8880-6f74-11ea-9277-4ed4a2b3af4c.png)

- [ ] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so I strongly recommend working on it

Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false
```

Reply:
We tried to do as you mentioned above, but it doesn't work on our local PC.
We don't know and understand what and how we should do for the my_map method.
So could you tell us what and how we should work our my_map method which we fixed?
<img width="650" alt="Screen Shot 2020-11-03 at 23 43 18" src="https://user-images.githubusercontent.com/53027195/97999337-85f9c680-1e2e-11eb-8953-7d1cc29b3209.png">

Reply:
I hope this [article ](https://docs.ruby-lang.org/en/2.6.0/Proc.html) might help you.

Reply:
We just fixed it, and it should work now. 
Actually, on our local PC, it's working now.
Thanks a lot.